arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/Fold.kt

```diff
@@ -224,17 +224,29 @@ public inline fun <Error, A> Raise<Error>.traced(
   @BuilderInference block: Raise<Error>.() -> A,
   trace: (trace: Trace, error: Error) -> Unit
 ): A {
-  val isOuterTraced = this is DefaultRaise && isTraced
-  val nested: DefaultRaise = if (isOuterTraced) this as DefaultRaise else DefaultRaise(true)
+  val nested = DefaultRaise(true)
   return try {
-    block.invoke(nested)
+    block(nested).also { nested.complete() }
   } catch (e: CancellationException) {
+    nested.complete()
     val r: Error = e.raisedOrRethrow(nested)
     trace(Trace(e), r)
-    if (isOuterTraced) throw e else raise(r)
+    // If our outer Raise happens to be traced
+    // Then we want the stack trace to match the inner one
+    try {
+      raise(r)
+    } catch (rethrown: CancellationException) {
+      throw rethrown.withCause(e)
+    }
```

Comment:
I'm not entirely sure I follow the changes here... When would `raise` throw `CancellationException`?

Reply:
The point here is to catch the `RaiseCancellationException` that `raise` will absolutely throw, and add our cause onto it to preserve the stacktrace information. Perhaps there's a clearer way to do that, but this is what I could come up with.